### ISSUE TITLE
(DDP) RoBERTa_large training with `torch.compile` results in OOM and other issues

### ISSUE THREAD

**Issue by 0x6b64** [pytorch/pytorch#93378]
### 🐛 Describe the bug

Hi,

I'm working with Jan 28 nightly build of PyTorch (nightly branch)
https://github.com/pytorch/pytorch/commit/5d6a4f697cac34d15262aad8afab096170d29ce1

RoBERTa architecture here: https://arxiv.org/pdf/1907.11692.pdf

The model definition && training scripts come from Fairseq
1. Trainer: https://github.com/facebookresearch/fairseq/blob/main/fairseq/trainer.py
2. Model def (RoBERTa Large): https://github.com/facebookresearch/fairseq/blob/main/fairseq/models/roberta/model.py#L34

The script was adapted by wrapping the `wrapped_model` in `trainer.py`

```
self._wrapped_model = torch.compile(self._wrapped_model)
```

I'm using the wiki text dataset: https://huggingface.co/datasets/wikitext

I'm working on an EC2 setup with `p4d.24xlarge` instance. 
Find instance specification here: https://aws.amazon.com/ec2/instance-types/p4/

These are the hyperparameter settings:

```bash
fairseq-train wikitext --adam-eps 1e-06 --arch roberta_large --attention-dropout 0.1 --clip-norm 0.0 --criterion masked_lm --distributed-backend nccl --distributed-no-spawn --dropout 0.1 --encoder-embed-dim 2048 --encoder-ffn-embed-dim 8192 --encoder-layers 24 --log-format simple --log-interval 10 --lr 0.0001 --lr-scheduler polynomial_decay --max-sentences 8 --max-update 500 --optimizer adam --sample-break-mode complete --skip-invalid-size-inputs-valid-test --task masked_lm --tokens-per-sample 512 --total-num-update 100 --update-freq 1 --weight-decay 0.01 --no-save --memory-efficient-fp16 --skip-invalid-size-inputs-valid-test --no-last-checkpoints
```

Note that to replicate this setup, you will have to rebuild Apex against the latest version of PT, I couldn't do this in the PyTorch nightly images because of CUDA missing dependencies on that container. Perhaps the PyTorch nightly devel container will have these dependencies (will validate later today).

Once that is done, you will have to rebuild Fairseq. 
Fairseq uses quite a few torchscript annotation and conditional code branching such as `if torch.jit.tracing: do_f`, and I've observed that dynamo symbolic converter throws this warning

```
[2023-01-28 17:07:39,262] torch._dynamo.symbolic_convert: [DEBUG] FAILED INLINING <code object forward_scriptable at 0x7fe3ac862f50, file "/fairseq/fairseq/models/transformer/transformer_encoder.py", line 173>
``` 
https://github.com/facebookresearch/fairseq/blob/main/fairseq/models/transformer/transformer_encoder.py#L173

> What is the guidance on using torchscript in conjunction to torch.compile? Is it recommended to not do that? 

However, this won't cause compilation to stop but down the road inductor will throw an error

```
  File "/opt/conda/lib/python3.9/site-packages/torch/_inductor/triton_ops/autotune.py", line 187, in run
    result = launcher(
  File "<string>", line 6, in launcher
ValueError: Pointer argument (at 0) cannot be accessed from Triton (cpu tensor?)
```

training works fine runs fine without torch.compile wrap. 

So, I'm not certain if this is an issue specifically with my setup but I can confirm that I can run a few of the dynamo benchmarks successfully. I'll try to get fairseq working on the nightly container

--

Anyways after removing the jit branches in code, the training continues but runs into OOM issues. The /fsx/pytorch-nightly/pytorch-2.0/pytorch/aten/src/ATen/cuda/CUDAGraph.cpp fails to allocate memory.

More questions. 

These warnings: 
/opt/conda/lib/python3.9/site-packages/torch/nn/functional.py:4872: UserWarning: This function is deprecated please rebuild your models with the public version of sdpa.
  warnings.warn("This function is deprecated please rebuild your models with the public version of sdpa.")

[2023-01-31 16:46:36,214] torch._inductor.lowering: [WARNING] using triton random, expect difference from eager

-> What is actionable for these? Specifically, for spda, what is the recommendation? Any particular library that should be used. What is the private (as opposed to public) version of SPDA?

Moreover, I've attached the minified scripts, but they don't reproduce the issue. Do they really help you as a domain specialist if the problem is not reproducible?

I've not tried running this without the DDP wrapper, but eventually I need to train this model/its variants on a large cluster, so even if the non DDP version works, thats not helpful. 

Moreover, all the artifacts pasted here are collected with `mpirun -np 1 <command>`, so use DDP with 1 rank. This is not relevant, but giving you the full picture. 

### Error logs

```
/opt/conda/lib/python3.9/site-packages/torch/cuda/graphs.py:82: UserWarning: The CUDA Graph is empty. This ususally means that the graph was attempted to be captured on wrong device or stream. (Triggered internally at /fsx/pytorch-nightly/pytorch-2.0/pytorch/aten/src/ATen/cuda/CUDAGraph.cpp:191.)
  super(CUDAGraph, self).capture_end()
2023-01-31 16:46:57 | WARNING | fairseq.trainer | |===========================================================================|
|                  PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 1            |        cudaMalloc retries: 1         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |  32019 MiB |  32019 MiB |  66253 MiB |  34234 MiB |
|       from large pool |  32007 MiB |  32007 MiB |  65929 MiB |  33922 MiB |
|       from small pool |     12 MiB |     12 MiB |    323 MiB |    311 MiB |
|---------------------------------------------------------------------------|
| Active memory         |  32019 MiB |  32019 MiB |  66253 MiB |  34234 MiB |
|       from large pool |  32007 MiB |  32007 MiB |  65929 MiB |  33922 MiB |
|       from small pool |     12 MiB |     12 MiB |    323 MiB |    311 MiB |
|---------------------------------------------------------------------------|
| GPU reserved memory   |  39034 MiB |  39034 MiB |  57400 MiB |  18366 MiB |
|       from large pool |  38856 MiB |  38856 MiB |  57160 MiB |  18304 MiB |
|       from small pool |    178 MiB |    178 MiB |    240 MiB |     62 MiB |
|---------------------------------------------------------------------------|
| Non-releasable memory |   4116 MiB |   4129 MiB |  15251 MiB |  11135 MiB |
|       from large pool |   3950 MiB |   3962 MiB |  14611 MiB |  10660 MiB |
|       from small pool |    165 MiB |    167 MiB |    640 MiB |    474 MiB |
|---------------------------------------------------------------------------|
| Allocations           |    2847    |    2847    |    4943    |    2096    |
|       from large pool |    1275    |    1275    |    2281    |    1006    |
|       from small pool |    1572    |    1572    |    2662    |    1090    |
|---------------------------------------------------------------------------|
| Active allocs         |    2847    |    2847    |    4943    |    2096    |
|       from large pool |    1275    |    1275    |    2281    |    1006    |
|       from small pool |    1572    |    1572    |    2662    |    1090    |
|---------------------------------------------------------------------------|
| GPU reserved segments |    1211    |    1211    |    1705    |     494    |
|       from large pool |    1122    |    1122    |    1585    |     463    |
|       from small pool |      89    |      89    |     120    |      31    |
|---------------------------------------------------------------------------|
| Non-releasable allocs |     549    |     549    |    1403    |     854    |
|       from large pool |     410    |     410    |     851    |     441    |
|       from small pool |     139    |     139    |     552    |     413    |
|---------------------------------------------------------------------------|
| Oversize allocations  |       0    |       0    |       0    |       0    |
|---------------------------------------------------------------------------|
| Oversize GPU segments |       0    |       0    |       0    |       0    |
|===========================================================================|
2023-01-31 16:46:57 | ERROR | fairseq.trainer | OOM during optimization, irrecoverable
Traceback (most recent call last):
  File "/opt/conda/bin/fairseq-train", line 8, in <module>
    sys.exit(cli_main())
  File "/fsx/roberta/fairseq/fairseq_cli/train.py", line 574, in cli_main
    distributed_utils.call_main(cfg, main)
  File "/fsx/roberta/fairseq/fairseq/distributed/utils.py", line 393, in call_main
    distributed_main(cfg.distributed_training.device_id, main, cfg, kwargs)
  File "/fsx/roberta/fairseq/fairseq/distributed/utils.py", line 366, in distributed_main
    main(cfg, **kwargs)
  File "/fsx/roberta/fairseq/fairseq_cli/train.py", line 205, in main
    valid_losses, should_stop = train(cfg, trainer, task, epoch_itr)
  File "/opt/conda/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/fsx/roberta/fairseq/fairseq_cli/train.py", line 331, in train
    log_output = trainer.train_step(samples)
  File "/opt/conda/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/fsx/roberta/fairseq/fairseq/trainer.py", line 1034, in train_step
    raise e
  File "/fsx/roberta/fairseq/fairseq/trainer.py", line 979, in train_step
    self.task.optimizer_step(
  File "/fsx/roberta/fairseq/fairseq/tasks/fairseq_task.py", line 545, in optimizer_step
    optimizer.step()
  File "/fsx/roberta/fairseq/fairseq/optim/fp16_optimizer.py", line 450, in step
    self.wrapped_optimizer.step(closure, groups=groups)
  File "/fsx/roberta/fairseq/fairseq/optim/fairseq_optimizer.py", line 135, in step
    self.optimizer.step(closure)
  File "/opt/conda/lib/python3.9/site-packages/torch/optim/optimizer.py", line 253, in wrapper
    out = func(*args, **kwargs)
  File "/fsx/roberta/fairseq/fairseq/optim/fused_adam.py", line 329, in step
    state["exp_avg_sq"] = torch.zeros_like(
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 64.00 MiB (GPU 0; 39.59 GiB total capacity; 31.27 GiB already allocated; 66.19 MiB free; 38.12 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

### Minified repro

Details too long to add in this comment. Repro details after dynamo/AOT are here: https://gist.github.com/0x6b64/966c2900fc7609d8bb9eaec22f8d6cc0

### Versions

```bash
Collecting environment information...
PyTorch version: 2.0.0a0+git5876d91
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.3
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:58:50)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-4.14.296-222.539.amzn2.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: NVIDIA A100-SXM4-40GB 

Nvidia driver version: 470.103.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.0
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.1.4
[pip3] functorch==1.14.0a0+408bcf1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.1
[pip3] torch==2.0.0a0+git5876d91
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.0.0a0+4699ef2
[pip3] torchdata==0.6.0a0+a1612ee
[pip3] torchmetrics==0.11.0
[pip3] torchrec-nightly==2023.1.29
[pip3] torchtext==0.15.0a0+f653dac
[pip3] torchvision==0.15.0a0+c35e8d5
[pip3] vector-quantize-pytorch==0.10.15
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.5.0                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.1.4                    pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] mkl-include               2023.0.0         h84fe81f_25396    conda-forge
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.1                    pypi_0    pypi
[conda] torch                     2.0.0a0+git5876d91          pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.0.0a0+4699ef2          pypi_0    pypi
[conda] torchdata                 0.6.0a0+a1612ee          pypi_0    pypi
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchrec-nightly          2023.1.29                pypi_0    pypi
[conda] torchtext                 0.15.0a0+f653dac          pypi_0    pypi
[conda] torchvision               0.15.0a0+c35e8d5          pypi_0    pypi
[conda] vector-quantize-pytorch   0.10.15                  pypi_0    pypi
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by wconstab** [comment]
I won't comment on the inductor+torchscript question, but for the OOM:
* DDP optimizer does split the graph into smaller pieces, and at the border of these subgraphs there will be additional tensors (activations, gradients) materialized relative to when compiling a whole program without splitting the graph
* cuda graphs requires additional memory (for parameters only, iiuc) for keeping a pinned input buffer alive for the parameters

You may experiment with disabling cudagraphs (`torch._inductor.config.triton.cudagraphs = False`) and disabling DDPOptimizer (`torch._dynamo.config.optimize_ddp = False`) to see if it gets you off the hook.  I'd also be curious to know how close to 100% the single-gpu model is without DDP and without cudagraphs, to get an intuition for whether this is a headroom issue or a bigger problem.

**Comment by 0x6b64** [comment]
Thanks for the response Will! Here is the observation with setting `cudagraphs = False` by adding 

```
import torch._inductor.config
torch._inductor.config.triton.cudagraphs = False
```

to Trainer.py https://github.com/facebookresearch/fairseq/blob/main/fairseq/trainer.py

```
/opt/conda/lib/python3.9/site-packages/torch/cuda/graphs.py:82: UserWarning: The CUDA Graph is empty. This ususally means that the graph was attempted to be captured on wrong device or stream. (Triggered internally at /fsx/pytorch-nightly/pytorch-2.0/pytorch/aten/src/ATen/cuda/CUDAGraph.cpp:191.)
  super(CUDAGraph, self).capture_end()
2023-01-31 17:54:01 | WARNING | fairseq.trainer | OOM: Ran out of memory with exception: CUDA out of memory. Tried to allocate 64.00 MiB (GPU 0; 39.59 GiB total capacity; 31.52 GiB already allocated; 36.19 MiB free; 38.15 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

Despite setting the config, the code flows through the cudagraphs.cpp && then throws "This ususally means that the graph was attempted to be captured on wrong device" and OOMs. 

--

With both `cudagraphs = False` && `optimize_ddp = False`

Same output.

**Comment by wconstab** [comment]
hmm, is there another cudagraph setting i am missing or that is part of your model or trainer definition?  It is surprising to me you still get any cudagraph warnings.

**Comment by 0x6b64** [comment]
Tried the same model on the Jan 25 nightly && that failed as well, though not with OOM.

```
[2023-02-03 00:42:45,070] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to multiple devices
./run_roberta_nccl_dynamo_local_torchrun.sh: line 56: 71240 Aborted                 fairseq-train /fsx/fairseq/data-bin/wikitext-103/ --adam-eps 1e-06 --arch roberta_large --attention-dropout 0.1 --clip-norm 0.0 --criterion masked_lm --distributed-backend nccl --distributed-no-spawn --dropout 0.1 --encoder-embed-dim 2048 --encoder-ffn-embed-dim 8192 --encoder-layers 24 --log-format simple --log-interval 10 --lr 0.0001 --lr-scheduler polynomial_decay --max-sentences 1 --max-update 500 --optimizer adam --sample-break-mode complete --skip-invalid-size-inputs-valid-test --task masked_lm --tokens-per-sample 512 --total-num-update 100 --update-freq 1 --weight-decay 0.01 --no-save --memory-efficient-fp16 --skip-invalid-size-inputs-valid-test --no-last-checkpoints --save-dir /fsx/karan/foobar
[2023-02-03 00:42:45,100] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to multiple devices
[2023-02-03 00:42:45,112] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to multiple devices
[2023-02-03 00:42:45,127] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to multiple devices
[2023-02-03 00:42:45,167] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to multiple devices
[2023-02-03 00:42:45,183] torch._inductor.compile_fx: [WARNING] skipping cudagraphs due to multiple devices
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71226 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71227 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71228 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71229 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71230 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71233 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 71237 closing signal SIGTERM
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 134) local_rank: 6 (pid: 71235) of binary: ./run_roberta_nccl_dynamo_local_torchrun.sh
Traceback (most recent call last):
  File "/opt/conda/bin/torchrun", line 33, in <module>
    sys.exit(load_entry_point('torch==2.0.0.dev20230124', 'console_scripts', 'torchrun')())
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/run.py", line 779, in main
    run(args)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/run.py", line 770, in run
    elastic_launch(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 250, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
```


Questions:
1. "skipping cudagraphs due to multiple devices" <- I'm guessing from the context that this should never be printed, so why not make it an error, than current warning.
2. `UserWarning: The CUDA Graph is empty. This ususally means that the graph was attempted to be captured on wrong device or stream. (Triggered internally at /opt/conda/conda-bld/pytorch_1674547949861/work/aten/src/ATen/cuda/CUDAGraph.cpp:191.)
  super(CUDAGraph, self).capture_end()` <- Is this also supposed to be a world stopping error?

**Comment by wconstab** [comment]
@0x6b64 Just a follow up on this:

Can you set `torch._dynamo.config.verbose = True` and rerun the case with cudagraph and ddp_optimizer disabled?

The goal is to collect a more detailed backtrace for the log above ("71240 Aborted" and "[WARNING] skipping cudagraphs")

Finally- I am wondering why you see "[WARNING] skipping cudagraphs" if you did set cudagraphs to be disabled in the config.  Can you confirm?

**Comment by 0x6b64** [comment]
Ack @wconstab - i'll prioritize collecting the observables for this today.

**Comment by chauhang** [comment]
@0x6b64 Can you please check if this issue resolved with latest PyTorch [2.0.1 release](https://github.com/pytorch/pytorch/releases/tag/v2.0.1)?

**Comment by williamwen42** [comment]
Closing due to inactivity - reopen if this is still an issue.
